core/src/main/scala/za/co/absa/spline/harvester/dispatcher/modelmapper/ModelMapperV12.scala

```diff
@@ -16,48 +16,110 @@
-    agentInfo = Some(automap(plan.agentInfo).to[v1_2.NameAndVersion]),
-    extraInfo = plan.extraInfo
+    systemInfo = toNameAndVersion(plan.systemInfo),
+    agentInfo = Some(toNameAndVersion(plan.agentInfo)),
+    extraInfo = toUntypedMap(plan.extraInfo).asNonEmptyOption
   ))
 
   def toOperations(operations: Operations): v1_2.Operations = v1_2.Operations(
-    write = automap(operations.write).to[v1_2.WriteOperation],
-    reads = operations.reads.map(automap(_).to[v1_2.ReadOperation]),
-    other = operations.other.map(automap(_).to[v1_2.DataOperation])
+    write = toWriteOperation(operations.write),
+    reads = operations.reads.map(toReadOperation).asNonEmptyOption,
+    other = operations.other.map(toDataOperation).asNonEmptyOption
   )
 
-  def toExpressions(expressions: Expressions): v1_2.Expressions = v1_2.Expressions(
-    functions = expressions.functions.map(automap(_).to[v1_2.FunctionalExpression]),
-    constants = expressions.constants.map(automap(_).to[v1_2.Literal])
+  def toWriteOperation(operation: WriteOperation): v1_2.WriteOperation = v1_2.WriteOperation(
+    outputSource = operation.outputSource,
+    append = operation.append,
+    id = operation.id,
+    name = Some(operation.name),
+    childIds = operation.childIds,
+    params = toUntypedMap(operation.params).asNonEmptyOption,
+    extra = toUntypedMap(operation.extra).asNonEmptyOption
   )
 
-  override def toDTO(event: ExecutionEvent): Option[v1_2.ExecutionEvent] = Some(automap(event).to[v1_2.ExecutionEvent])
+  def toReadOperation(operation: ReadOperation): v1_2.ReadOperation = v1_2.ReadOperation(
+    inputSources = operation.inputSources,
+    id = operation.id,
+    name = Some(operation.name),
+    output = operation.output.asNonEmptyOption,
+    params = toUntypedMap(operation.params).asNonEmptyOption,
+    extra = toUntypedMap(operation.extra).asNonEmptyOption
+  )
+
+  def toDataOperation(operation: DataOperation): v1_2.DataOperation = v1_2.DataOperation(
+    id = operation.id,
+    name = Some(operation.name),
+    childIds = operation.childIds.asNonEmptyOption,
+    output = operation.output.asNonEmptyOption,
+    params = toUntypedMap(operation.params).asNonEmptyOption,
+    extra = toUntypedMap(operation.extra).asNonEmptyOption
+  )
 
-  implicit def toAttrOrExprRef(o: Seq[AttrOrExprRef]): Option[Seq[v1_2.AttrOrExprRef]] = o.map {
+  def toAttribute(attribute: Attribute): v1_2.Attribute = v1_2.Attribute(
+    id = attribute.id,
+    dataType = attribute.dataType,
+    childRefs = attribute.childRefs.map(toAttrOrExprRef).asNonEmptyOption,
+    extra = toUntypedMap(attribute.extra).asNonEmptyOption,
+    name = attribute.name
+  )
+
+  def toAttrOrExprRef(o: AttrOrExprRef): v1_2.AttrOrExprRef = o match {
     case AttrRef(attrId) => v1_2.AttrOrExprRef(Some(attrId), None)
     case ExprRef(exprId) => v1_2.AttrOrExprRef(None, Some(exprId))
   }
+
+  val untypedMapConverter = new UntypedMapConverter(toAttrOrExprRef)
```

Comment:
Why do you need a shared instance of this class? From what I see it doesn't have any state, so can be just a function or an object, no?

Reply:
I changed it to function